### ISSUE TITLE
grpc_tools.protoc undefined symbol: _ZTVNSt7__cxx1119basic_istringstreamIcSt11char_traitsIcESaIcEEE

### ISSUE THREAD

**Issue by jsofri** [grpc/grpc#30218]
### What version of gRPC and what language are you using?
grpcio-tools 1.47.0
python

### What operating system (Linux, Windows,...) and version?
CentOS7.9  (linux) for ARM

### What runtime / compiler are you using (e.g. python version or version of gcc)
python3.6

### What did you do?
run `protoc` to compile a ".proto" file
`python3 -m grpc_tools.protoc -I. --grpc_python_out=. --python_out=. my.proto`

### What did you expect to see?
standard proto compilation, making auto generated scripts

### What did you see instead?
```
Traceback (most recent call last):
  File "/usr/lib64/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib64/python3.6/site-packages/grpc_tools/protoc.py", line 20, in <module>
    from grpc_tools import _protoc_compiler
ImportError: /usr/local/lib64/python3.6/site-packages/grpc_tools/_protoc_compiler.cpython-36m-aarch64-linux-gnu.so: undefined symbol: _ZTVNSt7__cxx1119basic_istringstreamIcSt11char_traitsIcESaIcEEE
```

### Anything else we should know about your project / environment?
from pip install
```
    Collecting grpcio-tools
      Downloading grpcio_tools-1.47.0-cp36-cp36m-manylinux_2_17_aarch64.whl (31.4 MB)
```

**Comment by gnossen** [comment]
Do you have a suitable version of `libstdc++` installed? Manylinux requires `libstdc++.so.6`. You should be able to check with the following command:

```
find /usr/lib -name 'libstdc++.so.6' 2>/dev/null
```

**Comment by gnossen** [comment]
There shouldn't be any system requirements besides those allowed by [the appropriate manylinux standard](https://peps.python.org/pep-0600/) (in your case, `manylinux_2_17`).

I think it's likely that the runtime linker is using this version of `libstdc++.so`. I suspect that you have a newish version of libstdc++ and there isn't a way that we can solve this issue on our end via manylinux. You should be able to verify this with the following command:

```
strings $(find /usr/lib -name 'libstdc++.so' 2>/dev/null | head -n1) | grep GLIBCXX
```

I'm not sure exactly which versions are affected, so it will be helpful to figure out known good version and known bad versions.

The missing symbol is `vtable for std::__cxx11::basic_istringstream<char, std::char_traits<char>, std::allocator<char> >`. In GCC 5, [a breaking ABI change was made to libstdc++ to enable copy-on-write semantics in `std::string`](https://stackoverflow.com/questions/34571583/understanding-gcc-5s-glibcxx-use-cxx11-abi-or-the-new-abi). However, none of the manylinux flavors yet support a compiler new enough to produce artifacts built against the new ABI. This is because the manylinux build environment is based on CentOS 7 and (somewhat ironically, considering you're using CentOS 7), CentOS 7 has [an outstanding bug](https://bugzilla.redhat.com/show_bug.cgi?id=1546704) that does not allow building with the new ABI.

Some ideas for workarounds:

- Try downgrading your `libstdc++`
- Try building from source.

**Comment by jsofri** [comment]
thank you

**Comment by beihai0xff** [comment]
same problem:
### Tools version
operating system: CentOS7.9 (linux) for ARM in docker
gcc version: gcc version 4.8.5 20150623 (Red Hat 4.8.5-44) (GCC) 
python version: 3.10.10
grpcio-tools: grpcio_tools-1.51.3-cp310-cp310-manylinux_2_17_aarch64.whl

### Error message
```bash
Traceback (most recent call last):
  File "/opt/smtx/python3/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/smtx/python3/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/root/.local/lib/python3.10/site-packages/grpc_tools/protoc.py", line 20, in <module>
    from grpc_tools import _protoc_compiler
ImportError: /root/.local/lib/python3.10/site-packages/grpc_tools/_protoc_compiler.cpython-310-aarch64-linux-gnu.so: undefined symbol: _ZTVNSt7__cxx1119basic_istringstreamIcSt11char_traitsIcESaIcEEE
```

### trace
```bash
[root@f4f0b132a96e]# ldd  /root/.local/lib/python3.10/site-packages/grpc_tools/_protoc_compiler.cpython-310-aarch64-linux-gnu.so
        linux-vdso.so.1 =>  (0x0000ffffbc144000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x0000ffffbbb4b000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x0000ffffbba24000)
        libm.so.6 => /lib64/libm.so.6 (0x0000ffffbb973000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x0000ffffbb942000)
        libc.so.6 => /lib64/libc.so.6 (0x0000ffffbb7bc000)
        /lib/ld-linux-aarch64.so.1 (0x0000ffffbc116000)
[root@f4f0b132a96e]# strings /usr/lib64/libstdc++.so.6 | grep GLIBCXX
  GLIBCXX_3.4
  GLIBCXX_3.4.1
  GLIBCXX_3.4.2
  GLIBCXX_3.4.3
  GLIBCXX_3.4.4
  GLIBCXX_3.4.5
  GLIBCXX_3.4.6
  GLIBCXX_3.4.7
  GLIBCXX_3.4.8
  GLIBCXX_3.4.9
  GLIBCXX_3.4.10
  GLIBCXX_3.4.11
  GLIBCXX_3.4.12
  GLIBCXX_3.4.13
  GLIBCXX_3.4.14
  GLIBCXX_3.4.15
  GLIBCXX_3.4.16
  GLIBCXX_3.4.17
  GLIBCXX_3.4.18
  GLIBCXX_3.4.19
  GLIBCXX_DEBUG_MESSAGE_LENGTH
```
### Others
if works fine on CentOS7.9 (linux) for x86_64, and GLIBCXX Version is same to arm

**Comment by eitkin-nvidia** [comment]
We encounter the same issue as well, and it blocks our suport for CentOS on arm. Could this please be nvestigated/addressed?

**Comment by cpungasc** [comment]
> Anyone found a solution to this issue, please? Thanks in advance!

`libstdc++.so.6.0.24` contains the symbol, it's available in`gcc-Version 7.3.1 20180712 (Red Hat 7.3.1-17) (GCC)` on `Amazon Linux`.

**Comment by cpungasc** [comment]
> > Anyone found a solution to this issue, please? Thanks in advance!
> 
> `libstdc++.so.6.0.24` contains the symbol, it's available in`gcc-Version 7.3.1 20180712 (Red Hat 7.3.1-17) (GCC)` on `Amazon Linux`.

for anyone travelling around these internet corners, installing from source `GCC 7.2.0` may be what you need, tested on `centos 7` and it worked.

**Comment by kalashnlkov** [comment]
reproduced on  grpcio-1.74.0-cp310-cp310-manylinux_2_17_aarch64.whl,    quay.io/pypa/manylinux2014_aarch64.
This problem is related to dual abi.
https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html

tldr. recompile grpcio-tool, grpcio under the same libcxxabi, or compile gcc with `--with-default-libstdcxx-abi=new` ( which is default)
